Stop duplicate messages being sent by `tedge-log-plugin` and `tedge-configuration-plugin` on file reload

Bravo555 opened the pull request:
## Proposed changes

This fix consists of 2 parts:

1. `NotifyStream` impl was changed to make use of a `notify-debouncer-full` crate to delay and debounce fs events. This should also impact other consumers of `NotifyStream`.
2. Watch parameters of `tedge-log-plugin` were changed to only listen to `Modified` events, no longer listening to `FileCreated` events.

This fix was confirmed to produce only 1 reload when using following editors:

- nano
- vim
- neovim

More details regarding both parts can be found in their respective commit messages.

## Types of changes

<!--
What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_
-->

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue

- #2451

## Checklist

<!--
_Put an `x` in the boxes that apply. You can also fill these out after
creating the PR. If you're unsure about any of them, don't hesitate to ask.
We're here to help! This is simply a reminder of what we are going to look for
before merging your code._
-->

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

reubenmiller commented on crates/extensions/tedge_log_manager/src/actor.rs:
```diff
@@ -222,7 +222,12 @@ impl LogManagerActor {
             FsWatchEvent::Modified(path) => path,
             FsWatchEvent::FileDeleted(path) => path,
-            FsWatchEvent::FileCreated(path) => path,
+            // if we created the file ealier and now modify it, keep only later modify. some editors
```
```suggestion
            // if we created the file earlier and now modify it, keep only later modify. some editors
```

Bravo555 replied:
I must've copied this comment on autopilot, because only now I realize that actually it doesn't talk about why `FileCreated` is disabled, but about something else. My bad. Will replace that with some more helpful comment.

didier-wenzek commented on crates/extensions/tedge_log_manager/src/actor.rs:
```diff
@@ -222,7 +222,13 @@ impl LogManagerActor {
             FsWatchEvent::Modified(path) => path,
             FsWatchEvent::FileDeleted(path) => path,
-            FsWatchEvent::FileCreated(path) => path,
+            // if we created the file earlier and now modify it, keep only later modify. some
+            // editors (e.g. vim) write to files by creating a temporary file, deleting old, and
+            // renaming temporary to previously deleted file. That way when we update, only 1 event
+            // is emitted.
+            //
+            // https://github.com/notify-rs/notify/issues/113#issuecomment-281836995
+            FsWatchEvent::FileCreated(_) => return Ok(()),
```
There is an issue with that. If the log plugin starts while its configuration is *not* present, then it starts assuming a default configuration. But then, as file creation events are ignored, there is no way to provide a log plugin configuration without restarting the plugin. I agree this is a corner case.

Bravo555 replied:
`FileCreated` means that a new file was created, but it still needs to be written to, so `FileCreated` is immediately followed by `FileModified` _most_ of the time, which the plugin picks up and correctly reloads.

I tested additional cases which result in successful reload:

- when the file is moved out and into the `/etc/tedge/plugins` directory (regardless if it was present when the plugin started) (the `FileCreated` event is not even emitted, because it's a `Rename` event, but `Modified` is still emitted)
- when plugin starts without the file and new one is created using `neovim` and by the shell (`echo "files = []" > tedge-log-plugin.toml`)
- when file is removed using `rm`

For all these the underlying library emits `EventKind::Modify(_)`.

reubenmiller commented:
> 2. Watch parameters of tedge-log-plugin were changed to only listen to Modified events, no longer listening to FileCreated events.

Are we also sure that this change does not affect replacing existing files using `mv` and `cp`? We have some [system test coverage](https://github.com/reubenmiller/thin-edge.io/blob/1db13f11810929ddddf5b5cacf3c75720e2bb7a4/tests/RobotFramework/tests/cumulocity/configuration/configuration_operation.robot#L185) for the configuration plugin, but none for the log manager. We should double check that we are not breaking any existing behaviour.

Also, shouldn't we also be doing the same handling for the configuration manager?

Bravo555 commented:
> Are we also sure that this change does not affect replacing existing files using `mv` and `cp`?

Moving files and creating files both also emit `Modified` event

> We have some [system test coverage](https://github.com/reubenmiller/thin-edge.io/blob/1db13f11810929ddddf5b5cacf3c75720e2bb7a4/tests/RobotFramework/tests/cumulocity/configuration/configuration_operation.robot#L185) for the configuration plugin, but none for the log manager. We should double check that we are not breaking any existing behaviour.

This is a good candidate for a Rust integration test, which I'll add for this PR.

> Also, shouldn't we also be doing the same handling for the configuration manager?

We probably should. I'll update the configuration manager as well.

Bravo555 commented on crates/common/tedge_utils/src/notify.rs:
```diff
@@ -57,59 +61,79 @@ impl NotifyStream {
+                    let mut debounced_events = debounced_events
+                        .into_iter()
+                        .filter_map(|notify_event| {
+                            let event = match notify_event.kind {
+                                EventKind::Access(access_kind) => match access_kind {
+                                    AccessKind::Close(access_mode) => match access_mode {
+                                        AccessMode::Any | AccessMode::Other | AccessMode::Write => {
+                                            Some(FsEvent::Modified)
+                                        }
```
This is the branch that allows us to only listen to `FsEvent::Modified`. The actual sequence returned by lower-level `notify-debouncer-full` is:

```
[crates/common/tedge_utils/src/notify.rs:77] &notify_event = DebouncedEvent {
    event: Event {
        kind: Create(
            File,
        ),
        paths: [
            "/etc/tedge/plugins/tedge-log-plugin.toml",
        ],
        attr:tracker: None,
        attr:flag: None,
        attr:info: None,
        attr:source: None,
    },
    time: Instant {
        tv_sec: 23846,
        tv_nsec: 112231843,
    },
}
[crates/common/tedge_utils/src/notify.rs:77] &notify_event = DebouncedEvent {
    event: Event {
        kind: Access(
            Close(
                Write,
            ),
        ),
        paths: [
            "/etc/tedge/plugins/tedge-log-plugin.toml",
        ],
        attr:tracker: None,
        attr:flag: None,
        attr:info: None,
        attr:source: None,
    },
    time: Instant {
        tv_sec: 23846,
        tv_nsec: 112263954,
    },
}
[crates/common/tedge_utils/src/notify.rs:77] &notify_event = DebouncedEvent {
    event: Event {
        kind: Remove(
            File,
        ),
        paths: [
            "/etc/tedge/plugins/tedge-log-plugin.toml~",
        ],
        attr:tracker: None,
        attr:flag: None,
        attr:info: None,
        attr:source: None,
    },
    time: Instant {
        tv_sec: 23846,
        tv_nsec: 112322524,
    },
}
```

when saving the file with nvim, or:

```
[crates/common/tedge_utils/src/notify.rs:77] &notify_event = DebouncedEvent {
    event: Event {
        kind: Create(
            File,
        ),
        paths: [
            "/etc/tedge/plugins/tedge-log-plugin.toml",
        ],
        attr:tracker: None,
        attr:flag: None,
        attr:info: None,
        attr:source: None,
    },
    time: Instant {
        tv_sec: 23953,
        tv_nsec: 87955441,
    },
}
[crates/common/tedge_utils/src/notify.rs:77] &notify_event = DebouncedEvent {
    event: Event {
        kind: Access(
            Close(
                Write,
            ),
        ),
        paths: [
            "/etc/tedge/plugins/tedge-log-plugin.toml",
        ],
        attr:tracker: None,
        attr:flag: None,
        attr:info: None,
        attr:source: None,
    },
    time: Instant {
        tv_sec: 23953,
        tv_nsec: 87977201,
    },
}
```

when file is copied. Picking up `AccessKind::Close(AccessMode::Write)` allows us to reload appropriately.

didier-wenzek reviewed (approved):
Approved. The code is clear and correct. Using `notify-debouncer-full` was a great idea!

didier-wenzek commented:
> Are we also sure that this change does not affect replacing existing files using `mv` and `cp`? 

This is working as expected.

albinsuresh reviewed (approved):
LGTM.

Bravo555 commented:
> This is a good candidate for a Rust integration test, which I'll add for this PR.

Instead of doing integration tests for both `tedge-log-plugin` and `tedge-configuration-plugin`, I've decided to only add 1 unit test: `notify::tests::modify_emitted_for_move_copy_create_delete` that checks if `FsEvent::Modified` is emitted when the file is modified, deleted, created, or renamed.

To add relevant tests for both plugins, I'd need to rewrite the plugins themselves a little bit, and we're currently more pressed for other stuff, so that should be sufficient for now.

albinsuresh commented on crates/common/tedge_utils/src/notify.rs:
```diff
@@ -267,21 +329,109 @@ mod tests {
+        assert_rx_stream(
+            [(
+                file.utf8_path().file_name().unwrap().to_string(),
+                vec![FsEvent::Modified],
```
Here I was expecting `FsEvent::FileCreated` event as well in addition to the `FsEvent::Modified` event, as the watcher logic above clearly matches `EventKind::Create` and maps it to `FsEvent::FileCreated`. So, what am I missing here?

Bravo555 replied:
In this test I wanted to ensure that `Modified` is emitted, testing for if `FileCreated` is emitted is done in other tests. `FileCreated` is emitted before `Modified`, but the `assert_rx_stream` just ignores them, it's only interested in events that are inside the vector.

albinsuresh commented on crates/common/tedge_utils/src/notify.rs:
```diff
@@ -57,59 +76,81 @@ impl NotifyStream {
-                                    AccessMode::Read | AccessMode::Execute => None,
+                                    CreateKind::Folder => Some(FsEvent::DirectoryCreated),
+                                },
+                                EventKind::Modify(modify_kind) => match modify_kind {
+                                    ModifyKind::Data(_)
+                                    | ModifyKind::Metadata(_)
+                                    | ModifyKind::Name(_)
+                                    | ModifyKind::Any
+                                    | ModifyKind::Other => Some(FsEvent::Modified),
```
```suggestion
                                EventKind::Modify(_) => Some(FsEvent::Modified),
```
Why not just that?

Bravo555 replied:
fixed

albinsuresh commented on crates/common/tedge_utils/src/notify.rs:
```diff
@@ -267,21 +329,109 @@ mod tests {
+        assert_rx_stream(
+            [(
+                file.utf8_path().file_name().unwrap().to_string(),
+                vec![FsEvent::Modified],
```
Here also, I was expecting the `FsEvent::FileDeleted` event right before the `FsEvent::Modified` event, as per the logic: `allow consumers to only monitor for `Modified` event if they want` implemented in that second `for` loop.

Bravo555 replied:
https://github.com/thin-edge/thin-edge.io/pull/2454#discussion_r1397172580

albinsuresh reviewed (commented):
Some queries to better understand the test logic.
